### PR TITLE
fix(discovery): Allows for negation of non-existent or unselected projects

### PR DESCRIPTION
**Pull request by rjo100**
Allows for negation of non-existent or unselected projects. Adds an empty value and essentially ignores the search term.

 Fixes SENTRY-YS6

### CODE REVIEW

File: src/sentry/search/events/filter.py [hunk 1]
```diff
@@ -961,10 +962,22 @@ def format_search_filter(term, params):
...
-            )
+        if missing:
+            if term.operator in EQUALITY_OPERATORS:
+                raise InvalidSearchQuery(
+                    f"Invalid query. Project(s) {oxfordize_list(missing)} do not exist or are not actively selected."
+                )
+            elif term.operator in INEQUALITY_OPERATORS:
+                # Set search filter to project name instead of id
+                term = SearchFilter(
+                    SearchKey("project.name"),
```
Comment (rjo100) [inline src/sentry/search/events/filter.py:973]: not sure if this is allowed. it's supposed to convert to `project_id` and convert the name to the id as well, but if it doesn't exist that won't work

Comment (barkbarkimashark) [reply src/sentry/search/events/filter.py:973]: I think searching on `project.name` results in a snuba query on `tags[project.name]` - not totally sure if SDKs or users send events with tag[project.name] values. In this case if the project slugs are missing and couldn't be resolved, could we just not add the filter to the condition?

**Comment by rjo100** [comment]
fixing this right now

File: src/sentry/search/events/filter.py (hunk 1 above)
Comment (barkbarkimashark) [reply src/sentry/search/events/filter.py:973]: Actually, there's probably a flaw in that approach. 

case when len(missing) == len(slugs):
  # couldn't resolve any of the project slugs to a project so we don't need to filter down on project.

case when len(missing) < len(slugs):
  # some slugs are un-resolved, but thats ok just let the code down below filter by project_id


Hmm, what if we

Comment (rjo100) [reply src/sentry/search/events/filter.py]: Yeah I was running into issues with the mixed case where some projects didn't exist/weren't selected and some were. I realized the problem was actually higher up the chain but regardless the filtering logic is a little restrictive down here.

Comment (rjo100) [reply src/sentry/search/events/filter.py]: Going to cut another PR to clean up the code once this is merged

File: src/sentry/search/snuba/executors.py [hunk 2]
```diff
@@ -164,19 +164,23 @@ def _convert_search_filters(
...
+            if conditions:
+                new_condition = conditions[0]
+            elif group_ids:
+                new_condition = convert_search_filter_to_snuba_query(
                     search_filter,
                     params={
                         "organization_id": organization_id,
                         "project_id": project_ids,
-                        "environment": environments,
+                        "environment_id": environments,
```
Comment (barkbarkimashark) [inline src/sentry/search/snuba/executors.py:178]: I think this has to be `environment`:
https://github.com/getsentry/sentry/blob/8e45ea7afa1b96de300f33fc92cd7f772639c5ff/src/sentry/search/events/filter.py#L573

Comment (rjo100) [reply src/sentry/search/snuba/executors.py:178]: oops didn't mean to change, copy paste error

File: tests/snuba/api/endpoints/test_organization_group_index.py [hunk 3]
```diff
@@ -449,6 +449,20 @@ def test_project(self):
         response = self.get_success_response(query=f"project:{project.slug}")
         assert len(response.data) == 1
 
+    def test_project_negation_invalid(self):
```
Comment (barkbarkimashark) [inline tests/snuba/api/endpoints/test_organization_group_index.py:452]: Could we add a testcase for `query="!project:[valid-project,invalid-project,also-invalid]".

Comment (rjo100) [reply tests/snuba/api/endpoints/test_organization_group_index.py:452]: yeah, I had one but deleted it haha

**Review by barkbarkimashark — APPROVED** [review]

**Pull request merged**